packages/components/multiselect/src/SelectAllOption.tsx

```diff
@@ -0,0 +1,91 @@
+import React from 'react';
+import { MultiselectOption } from './MultiselectOption';
+import { getMultiselectStyles } from './Multiselect.styles';
+
+type SelectAllOptionProps = {
+  childNodes: React.ReactNode;
+  iterateOverChildren: (
+    children: React.ReactNode,
+    filter: (child: React.ReactElement) => boolean,
+    callback: (child: React.ReactElement) => void,
+  ) => void;
+};
+
+const callChildEventHandler = (
+  child: React.ReactElement,
+  event: React.ChangeEvent<HTMLInputElement>,
+  checked: boolean,
+) => {
+  if (child.props?.isChecked !== checked && !child.props?.isDisabled) {
+    event.target.value = child.props?.value;
+    event.target.checked = checked;
+    child.props?.onSelectItem(event);
+  }
+};
+
+const areChildrenAllSelected = (
+  children: React.ReactNode,
+  filter: (child: React.ReactElement) => boolean,
+): boolean => {
+  let allSelected = true;
+  React.Children.forEach(children, (child) => {
+    // equal to (if (child == null || typeof child == 'string'))
+    if (!React.isValidElement(child)) return;
+    if (!filter(child)) {
+      allSelected = areChildrenAllSelected(child.props.children, filter);
+    } else {
+      if (child.props.isChecked) {
+        return;
+      }
+      allSelected = false;
+    }
+  });
+  return allSelected;
+};
+
+export const SelectAllOption = ({
+  childNodes,
+  iterateOverChildren,
+}: SelectAllOptionProps) => {
+  const styles = getMultiselectStyles();
+
+  const [allSelected, setAllSelected] = React.useState(
+    areChildrenAllSelected(
+      childNodes,
+      (child) => child.type === MultiselectOption,
+    ),
+  );
+
+  // checks if all children are true
+  React.useEffect(() => {
+    const allTrue = areChildrenAllSelected(
+      childNodes,
+      (child) => child.type === MultiselectOption,
+    );
+    setAllSelected(allTrue);
+  }, [childNodes]);
```

Comment:
I might be missing a obstacle here but I would try to avoid these kinds of state updates. You listen to a (transformed version) of an input variable and then set a state variable. I think that a `UseMemo` would be more appropriate since it saves you one render cycle. Here you first have one cycle with the new prop and then another one to update the state.

Reply:
If I'm not mistaken than this effet will get called when you change the `checked` state for a checkbox. This means that you don't need the state variable at all and can also save the call at line 73. By just updating the DOM nodes, this component rerenders and updates the memorized variable `allSelected` (instead of a controlled state)

Reply:
I might be wrong here because Multiselect is controlled but just wanted to raise the possibility of reducing complexity. Dunno if it's possible.

Reply:
i will have another look at it, but without the effect, it has some weird double-clicking required issues.

Reply:
I think it's also not a problem to have it this way. It's rather a micro-optimization that doesn't need to block this PR :)

Reply:
So i took another go at it and i don't need to call the method when initializing, its enough with the effect. In this case childNodes are props that change, not another state variable or anything inside. So useEffect is the correct hook over useCallback or useMemo.